### PR TITLE
Pass canonical URL on Safeframe metadata

### PR DESCRIPTION
**Pull request by bradfrizzell**

### CODE REVIEW

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js [hunk 1]
```diff
@@ -205,12 +206,17 @@ export class SafeframeHostApi {
           'readCookie': false,
           'writeCookie': false,
         }));
+    const ampDoc = Services.documentInfoForDoc(
+        this.baseInstance_.getAmpDoc());
+    const restricted = !!this.win_.document.querySelector(
+        "meta[name='referrer'][content='origin']");
```
Comment (jeffkaufman) [inline extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: Why are we only restricting on `origin` and not, say, `no-referrer`?  Can you add a comment describing your reasoning here?

(And maybe link to https://developer.mozilla.org/en-US/docs/Web/HTML/Element/meta and say we're trying to replicate that behavior, if that's what we're doing?)

Comment (bradfrizzell) [reply extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: Done

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js [hunk 2]
```diff
@@ -219,6 +221,25 @@ export class SafeframeHostApi {
...
+  /**
+   * Returns the canonical URL of the page, if the publisher allows
+   * it to be passed.
+   * @return {string|boolean}
+   */
+  maybeGetCanonicalUrl() {
+    // Don't allow for referrer policy same-origin,
+    // as Safeframe will always be a different origin.
+    // Don't allow for no-referrer.
+    if (!!this.win_.document.querySelector(
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: Let's save two query selectors either by changing this to:  querySelector("meta[name='referrer'][content='same-origin'], meta[name='referrer'][content='no-referrer']") 
OR
const metaReferrer = this.win_.document.querySelector("meta[name='referrer']");
if (metaReferrer && (metaReferrer.getAttribute('content') == 'same-origin' || metaReferrer.getAttribute('content') == 'no-referrer')) return false;

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js [hunk 3]
```diff
@@ -219,6 +221,25 @@ export class SafeframeHostApi {
     return attributes;
   }
 
+  /**
+   * Returns the canonical URL of the page, if the publisher allows
+   * it to be passed.
+   * @return {string|boolean}
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: Why not return empty string or undefined if not valid?  Boolean seems unnecessary

Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: @visibleForTesting

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js [hunk 4]
```diff
@@ -219,6 +221,25 @@ export class SafeframeHostApi {
...
+    // Don't allow for referrer policy same-origin,
+    // as Safeframe will always be a different origin.
+    // Don't allow for no-referrer.
+    if (!!this.win_.document.querySelector(
+        "meta[name='referrer'][content='same-origin']") ||
+        !!this.win_.document.querySelector(
+            "meta[name='referrer'][content='no-referrer']")) {
+      return false;
+    }
+    return Services.documentInfoForDoc(
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: this.baseInstance_. getAmpDoc().canonicalUrl

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js [hunk 5]
```diff
@@ -219,6 +221,25 @@ export class SafeframeHostApi {
...
+   * it to be passed.
+   * @return {string|boolean}
+   */
+  maybeGetCanonicalUrl() {
+    // Don't allow for referrer policy same-origin,
+    // as Safeframe will always be a different origin.
+    // Don't allow for no-referrer.
+    if (!!this.win_.document.querySelector(
+        "meta[name='referrer'][content='same-origin']") ||
+        !!this.win_.document.querySelector(
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: If origin, we should just return the origin of the canonical url (e.g. nytimes.com/foo?blah becomes nytimes.com)

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js (hunk 2 above)
Comment (bradfrizzell) [reply extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: done

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js (hunk 3 above)
Comment (bradfrizzell) [reply extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: done

Comment (bradfrizzell) [reply extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: done

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js (hunk 5 above)
Comment (bradfrizzell) [reply extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: done

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js [hunk 6]
```diff
@@ -219,6 +222,33 @@ export class SafeframeHostApi {
...
+   * @return {string|undefined}
+   * @visibleForTesting
+   */
+  maybeGetCanonicalUrl() {
+    // Don't allow for referrer policy same-origin,
+    // as Safeframe will always be a different origin.
+    // Don't allow for no-referrer.
+    const metaReferrer = this.win_.document.querySelector(
+        "meta[name='referrer']");
+    if (metaReferrer && (
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: Given you're checking against 3 values, let's do a switch:

const canonicalURL = ...
const metaReferrer = ...
if (!metaReferrer) return canonicalURL;
switch (metaReferrer.getAttribute('content')) {
    case 'same-origin':
    case 'no-referrer':
        return;
    case 'origin':
        return parseUrl(canonicalUrl).origin;
}
return canonicalUrl;

**Pull request merged**